sw/device/tests/alert_handler_lpg_clkoff_test.c

```diff
@@ -0,0 +1,743 @@
+};
+
+// This function will run after every device reset
+bool test_main(void) {
+  init_peripherals();
+
+  // A generic loop variable
+  int ii;
+  // To keep the test results
+  bool is_cause;
+
+  // Get the random wait times
+  uint32_t rnd_wait_times[ARRAYSIZE(kPeripherals)];
+  uint32_t max;
+  max = 0x1 << 10;
+  get_rand_words(&ibex, ARRAYSIZE(kPeripherals), rnd_wait_times, max);
+
+  // Enable flash access
+  flash_ctrl_testutils_default_region_access(&flash_ctrl,
+                                             /*rd_en*/ true,
+                                             /*prog_en*/ true,
+                                             /*erase_en*/ true,
+                                             /*scramble_en*/ false,
+                                             /*ecc_en*/ false,
+                                             /*he_en*/ false);
+
+  // Each test phase consists of ARRAYSIZE(kPeripherals) steps
+  // At every step, a specific peripheral is tested
+  size_t test_step_cnt = flash_ctrl_testutils_counter_get(kCounterTestSteps);
+  size_t test_phase = test_step_cnt / ARRAYSIZE(kPeripherals);
+  size_t peri_idx = test_step_cnt - (test_phase)*ARRAYSIZE(kPeripherals);
+
+  // At the beginning of the simulation (Verilator, VCS),
+  // the content of the flash might be all-zeros, and thus,
+  // the NVM counter's inital value might be 256.
+  // In that case, the NVM counter should be initialized to zero by filling
+  // its flash region with non-zero values.
+  // On FPGA, this code will be skipped
+  if (test_step_cnt == 256) {
+    flash_ctrl_testutils_counter_init_zero(&flash_ctrl, kCounterTestSteps);
+    test_step_cnt = flash_ctrl_testutils_counter_get(kCounterTestSteps);
+    test_phase = test_step_cnt / ARRAYSIZE(kPeripherals);
+    peri_idx = test_step_cnt - (test_phase)*ARRAYSIZE(kPeripherals);
+  }
+
+  /* TEST PHASE #0: NEGATIVE TESTING
+
+    The timeout value is set to 2
+    All of the alerts should trigger ping_timeout_alert
+    if they are not clock_gated
+    TEST: for_each_peripheral
+      1- Disable the clock
+      2- Enable the corresponding alerts
+      3- Wait long enough
+      4- Confirm that ping_timeout has not been triggerred
+      5- Reset the the device to test the next peripheral
+  */
+  while (test_step_cnt < 1 * ARRAYSIZE(kPeripherals)) {
+    // LOG_INFO("Test phase = %d, Periphereal idx = %d, test_step_cnt = %d",
+    // test_phase, peri_idx, test_step_cnt);
+    busy_spin_micros(rnd_wait_times[peri_idx]);
+
+    // Disable the clock of the peripheral
+    if (kPeripherals[peri_idx].is_hintable) {
+      CHECK_DIF_OK(dif_clkmgr_hintable_clock_set_hint(
+          &clkmgr, kPeripherals[peri_idx].clk_index, kDifToggleDisabled));
+    } else {
+      CHECK_DIF_OK(dif_clkmgr_gateable_clock_set_enabled(
+          &clkmgr, kPeripherals[peri_idx].clk_index, kDifToggleDisabled));
+    }
+
+    alert_handler_config_single_peripheral(&kPeripherals[peri_idx], 2);
+
+    // wait enough
```

Comment:
Similar to https://github.com/lowRISC/opentitan/blob/master/hw/ip_templates/alert_handler/rtl/alert_handler_ping_timer.sv#L160

I use (NUM_ALERTS*2) because the FSM waits in AlertWaitSt and EscWaitSt for each disabled peripheral, until it hits the enabled peripheral. 

I use MARGIN_OF_SAFETY=4, worked well on FPGA and in Verilator